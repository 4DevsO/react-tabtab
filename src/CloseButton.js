// @flow
import React from 'react';
import Close from 'react-icons/lib/md/clear';
import styled from 'styled-components';

const CloseWrapper = styled.div`
  display: inline-block;
  color: #777;
  margin-left: 3px;
  padding-bottom: 2px;
  vertical-align: middle;
  transition: background-color .2s cubic-bezier(.645, .045, .355, 1);
  &:hover {
    color: black;
    background-color: #eee;
    cursor: pointer;
  }
`;

type Props = {
  handleDelete: (event: any) => void
};

export default class CloseButton extends React.Component<Props> {
  render() {
    return (
      <CloseWrapper onClick={this.props.handleDelete}>
        <Close/>        
      </CloseWrapper>
    );
  }
}
