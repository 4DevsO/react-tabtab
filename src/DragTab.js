import React from 'react';
import {SortableElement} from 'react-sortable-hoc';
import Tab from './Tab';

const DragTabElement = SortableElement(({children, ...props}) => {
  return (
    <Tab index={props.tabIndex} {...props}>
      {children}
    </Tab>
  )
});

class DragTab extends React.Component {
  render() {
    const {children, ...props} = this.props;
    return (
      <DragTabElement {...props}>
        {children}
      </DragTabElement>
    )
  }
}

export default DragTab;