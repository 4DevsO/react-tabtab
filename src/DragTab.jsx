import React from 'react';
import ReactDOM from 'react-dom';
import { DragSource, DropTarget } from 'react-dnd';
import ItemTypes from './ItemTypes';
import Tab from './Tab';

const tabSource = {
  beginDrag: function(props) {
    props.beginDrag();
    return {
      tabKey: props.tabKey
    }
  }
}

// reference from https://github.com/gaearon/react-dnd/tree/master/examples/04%20Sortable/Simple
const tabTarget = {
  hover(props, monitor, component) {
    const dragIndex = monitor.getItem().tabKey;
    const hoverIndex = props.tabKey;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = ReactDOM.findDOMNode(component).getBoundingClientRect();

    // Get horizontal middle
    const hoverMiddleX = (hoverBoundingRect.right - hoverBoundingRect.left) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    const hoverClientX = hoverBoundingRect.right - clientOffset.x;

    // Dragging left
    if (dragIndex < hoverIndex && hoverClientX < hoverMiddleX) {
      return;
    }

    // Dragging right
    if (dragIndex > hoverIndex && hoverClientX > hoverMiddleX) {
      return;
    }

    props.moveTab(dragIndex, hoverIndex);
    // change the dragging tab index
    monitor.getItem().tabKey = hoverIndex;
  }
};

@DropTarget(ItemTypes.TAB, tabTarget, connect => ({
  connectDropTarget: connect.dropTarget()
}))
@DragSource(ItemTypes.TAB, tabSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging()
}))

export default class DragTab extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {

    const {connectDragSource, connectDropTarget } = this.props;

    return connectDragSource(connectDropTarget(
      <span>
        <Tab {...this.props}/>
      </span>
    ));
  }
}