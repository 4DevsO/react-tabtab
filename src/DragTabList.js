import React from 'react';
import {SortableContainer} from 'react-sortable-hoc';
import TabList from './TabList';

const DragTabContainer = SortableContainer(({children, ...props}) => {
  return (
    <TabList {...props}>
      {children}
    </TabList>
  );
});

export default class DragTabList extends React.Component {
  constructor(props) {
    super(props);
    this.onSortEnd = this.onSortEnd.bind(this);
  }

  onSortEnd({oldIndex, newIndex}) {
    const {activeIndex, handleActiveIndex, handleTabSequence} = this.props;
    if (oldIndex === newIndex) {
      if (activeIndex !== oldIndex) {
        handleActiveIndex(oldIndex);
      }
    } else {
      console.log('switch: ', oldIndex, newIndex)
      handleTabSequence({oldIndex, newIndex})
    }
  }

  render() {
    const {children, ...props} = this.props; // eslint-disable-line
    return (
      <DragTabContainer onSortEnd={this.onSortEnd}
                        axis='x'
                        lockAxis='x'
                        {...props}>
        {children}
      </DragTabContainer>
    );
  }
}

DragTabList.displayName = 'DragTabList';
