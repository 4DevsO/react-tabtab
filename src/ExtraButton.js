import React from 'react';
import styled from 'styled-components';

const Wrapper = styled.div`
  float: right;
  border: 1px solid #eee;
  border-radius: 2px;
  padding: 3px;
  margin-top: 10px;
  margin-left: 2px;
  display: inline-block;
  color: #777;
  vertical-align: middle;
  transition: background-color .2s cubic-bezier(.645, .045, .355, 1);
  &:hover {
    color: black;
    cursor: pointer;
  }
`;

export default class ExtraButton extends React.Component {
  render() {
    return (
      <Wrapper onClick={this.props.onClick}>
        <div>
          {this.props.children}          
        </div>
      </Wrapper>
    );
  }
}
