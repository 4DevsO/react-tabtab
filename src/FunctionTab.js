// import React from 'react';
// import classNames from 'classnames';
// import Tappable from 'react-tappable';

// export default class FunctionTab extends React.Component {
//   constructor(props) {
//     super(props);
//     this.clickTab = this.clickTab.bind(this);
//   }

//   clickTab() {
//     this.props.handleTabClick();
//   }

//   render() {
//     var tabClass;

//     tabClass = classNames(this.props.style + 'tab', 'add');
    
//     return (
//       <span>
//         <li className={tabClass} onClick={this.clickTab}>
//           {this.props.title}
//         </li>
//       </span>
//     );
//   }
// };