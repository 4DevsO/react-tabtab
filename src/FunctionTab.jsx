import React from 'react';
import classNames from 'classnames';
import Tappable from 'react-tappable';

export default class FunctionTab extends React.Component {
  constructor(props) {
    super(props);
    this.handleTabDeleteButton = this.handleTabDeleteButton.bind(this);
    this.clickTab = this.clickTab.bind(this);
  }

  handleTabDeleteButton(e) {
    e.stopPropagation(); //prevent trigger clickTab function
    this.props.handleTabDeleteButton(this.props.tabKey);
  }

  clickTab() {
    this.props.handleTabClick(this.props.tabKey);
  }

  render() {
    var tabClass;

    tabClass = classNames(this.props.style + 'tab', 'add');
    // only show the delete button when it's active

    return (
      <span>
        <li className={tabClass} onClick={this.clickTab}>
          {this.props.title}
        </li>
      </span>
    );
  }
};