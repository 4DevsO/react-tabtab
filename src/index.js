// @flow
import Tabs from './Tabs';
import TabList, {TabListStyle, ActionButtonStyle} from './TabList';
import Tab, {TabStyle} from './Tab';
import DragTabList from './DragTabList';
import DragTab from './DragTab';
import PanelList from './PanelList';
import Panel, {PanelStyle} from './Panel';
import ExtraButton from './ExtraButton';

const styled = {TabListStyle, ActionButtonStyle, TabStyle, PanelStyle};
const defaultOutput = {
  Tabs,
  TabList,
  Tab,
  DragTabList,
  DragTab,
  PanelList,
  Panel,
  ExtraButton,
  styled
}

export default defaultOutput;