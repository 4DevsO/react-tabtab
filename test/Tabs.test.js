import React from 'react';
import {Tabs, DragTabList, DragTab, TabList, Tab, PanelList, Panel} from '../src';
import {shallow, mount} from 'enzyme';

const normalComponent = (
  <Tabs>
    <TabList>
      <Tab>Tab1</Tab>
      <Tab>Tab2</Tab>
    </TabList>
    <PanelList>
      <Panel>Content1</Panel>
      <Panel>Content2</Panel>
    </PanelList>
  </Tabs>
);

const dragComponent = (
  <Tabs>
    <DragTabList>
      <DragTab>DragTab1</DragTab>
      <DragTab>DragTab2</DragTab>
    </DragTabList>
    <PanelList>
      <Panel>Content1</Panel>
      <Panel>Content2</Panel>
    </PanelList>
  </Tabs>
);

describe('render Tabs', () => {
  it('normal tabs', () => {
    const component = mount(normalComponent);
    expect(component.html()).toMatchSnapshot();
  });
  it('normal drag tabs', () => {
    const component = mount(dragComponent);
    expect(component.html()).toMatchSnapshot();
  });
});

describe('props', () => {
  it('defaultIndex', () => {
    const defaultIndex = 1;
    const component = shallow(
      <Tabs defaultIndex={defaultIndex}>
        <TabList>
          <Tab>Tab1</Tab>
          <Tab>Tab2</Tab>
        </TabList>
        <PanelList>
          <Panel>Content1</Panel>
          <Panel>Content2</Panel>
        </PanelList>
      </Tabs>
    );
    expect(component.state().activeIndex).toEqual(defaultIndex);
  })
})
