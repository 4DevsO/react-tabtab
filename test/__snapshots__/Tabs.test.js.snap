// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Tabs normal drag tabs 1`] = `"<div><div><div class=\\"sc-bdVaJa hjajR\\"><div><div class=\\"sc-fjdhpX gAbfpE\\"><svg viewBox=\\"0 0 40 40\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\" style=\\"vertical-align: middle;\\"><g><path d=\\"m25.7 12.3l-7.7 7.7 7.7 7.7-2.3 2.3-10-10 10-10z\\"></path></g></svg></div><div class=\\"sc-fjdhpX gYthMW\\"><svg viewBox=\\"0 0 40 40\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\" style=\\"vertical-align: middle;\\"><g><path d=\\"m16.6 10l10 10-10 10-2.3-2.3 7.7-7.7-7.7-7.7z\\"></path></g></svg></div></div><div class=\\"sc-bwzfXH fvdpZL\\"><div class=\\"sc-htpNat ezotGS\\"><div class=\\"sc-EHOje NvgmN\\">DragTab1</div><div class=\\"sc-EHOje bxODLe\\">DragTab2</div></div></div></div></div><div><div class=\\"sc-bZQynM iTruvD\\">Content1</div></div></div>"`;

exports[`render Tabs normal tabs 1`] = `"<div><div><div class=\\"sc-bdVaJa hjajR\\"><div><div class=\\"sc-gZMcBi kGTUtH\\"><svg viewBox=\\"0 0 40 40\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\" style=\\"vertical-align: middle;\\"><g><path d=\\"m25.7 12.3l-7.7 7.7 7.7 7.7-2.3 2.3-10-10 10-10z\\"></path></g></svg></div><div class=\\"sc-gZMcBi fPUvwG\\"><svg viewBox=\\"0 0 40 40\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\" style=\\"vertical-align: middle;\\"><g><path d=\\"m16.6 10l10 10-10 10-2.3-2.3 7.7-7.7-7.7-7.7z\\"></path></g></svg></div></div><div class=\\"sc-bwzfXH fvdpZL\\"><div class=\\"sc-htpNat ezotGS\\"><div class=\\"sc-EHOje NvgmN\\">Tab1</div><div class=\\"sc-EHOje bxODLe\\">Tab2</div></div></div></div></div><div><div class=\\"sc-bZQynM iTruvD\\">Content1</div></div></div>"`;
